pkg/objstore/swift/swift.go

```diff
@@ -9,69 +9,188 @@ import (
+		ChunkSize:              DefaultConfig.ChunkSize,
+		SegmentContainerName:   os.Getenv("SWIFT_SEGMENTS_CONTAINER_NAME"),
+		Retries:                c.Retries,
+		ConnectTimeout:         model.Duration(c.ConnectTimeout),
+		Timeout:                model.Duration(c.Timeout),
+		UseDynamicLargeObjects: false,
+	}
+	if os.Getenv("SWIFT_CHUNK_SIZE") != "" {
+		var err error
+		config.ChunkSize, err = strconv.ParseInt(os.Getenv("SWIFT_CHUNK_SIZE"), 10, 64)
+		if err != nil {
+			return nil, errors.Wrap(err, "swift parsing chunk size")
+		}
+	}
+	if strings.ToLower(os.Getenv("SWIFT_USE_DYNAMIC_LARGE_OBJECTS")) == "true" {
+		config.UseDynamicLargeObjects = true
+	}
+	return &config, nil
+}
+
+func connectionFromConfig(sc *Config) *swift.Connection {
+	connection := swift.Connection{
+		Domain:         sc.DomainName,
+		DomainId:       sc.DomainId,
+		UserName:       sc.Username,
+		UserId:         sc.UserId,
+		ApiKey:         sc.Password,
+		AuthUrl:        sc.AuthUrl,
+		Retries:        sc.Retries,
+		Region:         sc.RegionName,
+		AuthVersion:    sc.AuthVersion,
+		Tenant:         sc.ProjectName,
+		TenantId:       sc.ProjectID,
+		TenantDomain:   sc.ProjectDomainName,
+		TenantDomainId: sc.ProjectDomainID,
+		ConnectTimeout: time.Duration(sc.ConnectTimeout),
+		Timeout:        time.Duration(sc.Timeout),
+	}
+	return &connection
 }
 
 type Container struct {
-	logger log.Logger
-	client *gophercloud.ServiceClient
-	name   string
+	logger                 log.Logger
+	name                   string
+	connection             *swift.Connection
+	chunkSize              int64
+	useDynamicLargeObjects bool
+	segmentsContainer      string
 }
 
 func NewContainer(logger log.Logger, conf []byte) (*Container, error) {
 	sc, err := parseConfig(conf)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "swift parse config")
 	}
+	return NewContainerFromConfig(logger, sc, false)
+}
 
-	provider, err := openstack.AuthenticatedClient(authOptsFromConfig(sc))
-	if err != nil {
-		return nil, err
+func ensureContainer(connection *swift.Connection, name string, createIfNotExist bool) error {
+	if _, _, err := connection.Container(name); err != nil {
+		if err != swift.ContainerNotFound {
+			return errors.Wrapf(err, "swift verify container %s", name)
+		}
+		if !createIfNotExist {
+			return fmt.Errorf("unable to find the expected container %s", name)
+		}
+		if err = connection.ContainerCreate(name, swift.Headers{}); err != nil {
+			return errors.Wrapf(err, "create container %s", name)
+		}
+		return nil
 	}
+	return nil
+}
 
-	client, err := openstack.NewObjectStorageV1(provider, gophercloud.EndpointOpts{
-		Region: sc.RegionName,
-	})
-	if err != nil {
+func NewContainerFromConfig(logger log.Logger, sc *Config, createContainer bool) (*Container, error) {
+	connection := connectionFromConfig(sc)
+	if err := connection.Authenticate(); err != nil {
+		return nil, errors.Wrap(err, "swift authentication")
```

Comment:
Normally we don't need to add `swift` to wrap. In the same way, we don't need to add `NewContainerFromConfig swift...`. This is because in Thanos we wrap "whatever we do now" with context, not "where we do it"

Reply:
Right, I removed the swift reference from all the error wrappings.